feat: Add Task support for BatchExportProcessor and PeriodicExportingMetricReader for single-threaded WebAssembly

jeromelaban opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-dotnet/issues/2816
Design discussion issue: Partially based on https://github.com/open-telemetry/opentelemetry-dotnet/issues/5778 and https://github.com/open-telemetry/opentelemetry-dotnet/pull/5838

## Changes

The change introduces updates to `BatchExportProcessor` and `PeriodicExportingMetricReader` in order to switch between Thread and Task implementations to support .NET's WebAssembly single threaded mode (used by Blazor, Uno Platform and others).

Based on https://github.com/open-telemetry/opentelemetry-dotnet/pull/5838#pullrequestreview-2479533689, the threads mode is automatically enabled when multi-threading is available. Tasks are enabled when WebAssembly is detected, based on the `ThreadingHelper` class. 

Note that there is no public API to change between thread and tasks.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

jeromelaban committed:
feat: Add optional Task support for BatchExportProcessor and PeriodicExportingMetricReader

This change adds the optional support for threads for BatchExportProcessor and PeriodicExportingMetricReader, in order to support .NET's WebAssembly single threaded mode.

jeromelaban committed:
chore: Adjust parameters ordering to avoid unnamed parameters breaking change

jeromelaban commented:
To whoever approved the workflows run, thank you :) 

I've made adjustments to avoid parameters reordering breaking changes which should get the build go further.

jeromelaban committed:
chore: Add processor dispose validation test

jeromelaban commented:
The tests revealed a possible race condition in the disposing/shutdown ordering of the processor resulting in missing exports, for the task implementation.

jeromelaban committed:
chore: Remove early exit of the exporter

Allow for slow thread startup to process exports if the processor is shut down quickly

jeromelaban commented:
Thanks for the run, again! 

Looks like there's a reported test failure, but without any failed tests. Feels like a fluke, but it may be something known to this project. Would a rerun help? Thanks!

jeromelaban commented:
Thanks @cijothomas, it helped :)

jeromelaban committed:
chore: Adjust exit condition for ExporterProcAsync

This will avoid exiting on an exceptional flow on Task.WhenAny

jeromelaban committed:
chore: Reorder test parameters in xml doc

jeromelaban committed:
chore: Use Task.WhenAny's output to avoid a double await

Also enable tests for PeriodicExportingMetricReaderTaskWorker to improve coverage

jeromelaban commented on src/OpenTelemetry/Metrics/Reader/PeriodicExportingMetricReader.cs:
```diff
@@ -104,41 +103,25 @@ protected override void Dispose(bool disposing)
+            this,
+            this.ExportIntervalMilliseconds,
+            this.ExportTimeoutMilliseconds);
     }
```
Fixed.

jeromelaban commented on src/OpenTelemetry/Internal/BatchExportTaskWorker.cs:
```diff
@@ -207,7 +205,7 @@ private async Task ExporterProcAsync()
         try
         {
-            while (!cancellationToken.IsCancellationRequested && !this.isShutdownRequested)
+            while (true)
```
Indeed. The original intent was to let Task.WhenAny's parameters throw when canceled, but it's better to avoid this exceptional flow.  Thanks!

jeromelaban committed:
chore: Adjust volatile for potential cross-thread access

In theory, there should not be multiple threads access to this field, but this won't hurt to add this qualifier.

jeromelaban committed:
chore: Remove isShutdownRequested

It's a duplicate of cancellationToken.IsCancellationRequested

jeromelaban committed:
chore; Explain the reason for the disabled warning

jeromelaban committed:
chore: Remove unused namespace

jeromelaban commented on src/OpenTelemetry/.publicApi/Stable/PublicAPI.Shipped.txt:
Indeed, that's what I mentioned in the original discussion.

The problem here is that adding a parameter to multiple classes creates this error:

```
warning RS0026: Symbol 'BatchExportProcessor' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'.
```

Which forces to change the original ctor to remove the optional values and move them to the new constructor in order to really be backwards compatible.

Changing default values is not an actual breaking change. Default values are captured by the compiled code at compile time, but it does not impact method binding to a newer version which does not have the optional parameters. This PR does not change the default values for the new ctor, which preserves the original behavior.

jeromelaban replied:
The shipped file still does only contain the shipped APIs. It's the same signature, minus the optional parameters. The new signatures are part of the unshipped file (with the `useThreads` parameter).

I don't think it's possible to ignore the warning either, mostly because the two signature will be ambiguous if both have optional parameters, for instance:

```csharp
new BatchExportProcessor(exporter)
```

will match both:

```csharp
    public PeriodicExportingMetricReader(
        BaseExporter<Metric> exporter,
        int exportIntervalMilliseconds = DefaultExportIntervalMilliseconds,
        int exportTimeoutMilliseconds = DefaultExportTimeoutMilliseconds)
```

and

```csharp
    public PeriodicExportingMetricReader(
        BaseExporter<Metric> exporter,
        int exportIntervalMilliseconds = DefaultExportIntervalMilliseconds,
        int exportTimeoutMilliseconds = DefaultExportTimeoutMilliseconds,
        bool useThreads = true)
```

Another solution could be to add a static member named `PeriodicExportingMetricReader.UseThreads` and `BatchLogRecordExportProcessor.UseThreads` that can be optionally set before anything else, which could solve this overall optional parameters issue.

martincostello replied:
I've looked at the diff again, and there is definitely a binary-breaking change made in this diff.

You cannot remove the optional parameters - doing so is only non-breaking at _compile_ time. It will create `MissingMethodException` at runtime as already compiled code (e.g. dependencies of dependencies which aren't recompiled when the dependency is updated) as they will try to find the signature already in the shipped code and will fail to bind to it.

This change need to add the new parameter you want to add without changing any signatures in the `Shipped` file.

jeromelaban replied:
I've validated again my assumptions about optional parameters (just in case I missed an new runtime behavior :) ) using a small sample project (see at the end), and optional parameters are not binary breaking if removed, as long as the original method with the right signature is still present.

Where there may be runtime side effects is if the new method behaves differently when automatically selected new version of the ctor. It's not the case here, since the parameters are the same, and the defaults are the same when chaining the new ctor.

Nonetheless, I understand that it's against the philosophy of the shipped APIs to change its contents, so would you be open to the static property API? Still, I'm generally against it because it can break testing in a very challenging ways when tests are running multi-threaded.

There could also be the ability to use a static method such as `PeriodicExportingMetricReader.CreateWithTasks()` which would use an private ctor with the `useThreads` parameter.

Otherwise, would you have another suggestion?

---

[6379-testbreaking.zip](https://github.com/user-attachments/files/21348338/6379-testbreaking.zip). Compile once with the version with one ctor, then in-place replace the dll's binary with the new ctor, the executed code still runs without any issues (and still call the original signature).

jeromelaban committed:
chore: Move to BatchLogRecordExportProcessorOptions and PeriodicExportingMetricReaderOptions

This avoids introducing a potentially ambiguous constructor, restores the original shipped API signatures.

martincostello commented on src/OpenTelemetry/BatchExportProcessorOptions.cs:
```diff
@@ -29,4 +29,9 @@ public class BatchExportProcessorOptions<T>
     public int MaxExportBatchSize { get; set; } = BatchExportProcessor<T>.DefaultMaxExportBatchSize;
+
+    /// <summary>
+    /// Gets or sets a value indicating whether to use threads. Enables the use of <see cref="Thread" /> when true, <see cref="Task"/> when false. The default value is true.
```
```suggestion
    /// Gets or sets a value indicating whether to use threads. Enables the use of <see cref="Thread" /> when <see langword="true"/>; otherwise <see cref="Task"/> is used. The default value is <see langword="true"/>.
```

martincostello commented on src/OpenTelemetry/Metrics/Reader/PeriodicExportingMetricReader.cs:
```diff
@@ -33,8 +32,32 @@ public PeriodicExportingMetricReader(
     {
+        Guard.ThrowIfNull(options);
+
+#pragma warning disable CA1062 // Validate arguments of public methods - needed for netstandard2.1
+        var exportIntervalMilliseconds = options!.ExportIntervalMilliseconds ?? DefaultExportIntervalMilliseconds;
+#pragma warning restore CA1062 // Validate arguments of public methods - needed for netstandard2.1
```
Would using `?` instead of `!` make it happy without needing the `#pragma`?

jeromelaban replied:
It does, though it needs to be specified in all uses of `options`. I've mainly used this because it's the same pattern used all around the library, but null propagation is also fine :) Will adjust.

martincostello commented on src/OpenTelemetry/Metrics/Reader/PeriodicExportingMetricReader.cs:
```diff
@@ -46,45 +69,30 @@ public PeriodicExportingMetricReader(
         }
 
-        return result;
+        var sw = Stopwatch.StartNew();
+        var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;
```
This code looks redundant now. There's no task happening between the stopwatch starting and timeout calculation now, so it should just be `this.exporter.Shutdown(timeout) && result`?

martincostello commented on src/OpenTelemetry/Metrics/Reader/PeriodicExportingMetricReader.cs:
```diff
@@ -104,41 +111,28 @@ protected override void Dispose(bool disposing)
-        int timeout;
-        var triggers = new WaitHandle[] { this.exportTrigger, this.shutdownTrigger };
-        var sw = Stopwatch.StartNew();
+    // The pragma is required by the fact that this method is compiled on both .NET Framework and .NET Core, and the CA1859 warning is only relevant for .NET Framework.
+    // The warning suggests changing the return type to PeriodicExportingMetricReaderThreadWorker for improved performance, but we want to keep the method signature consistent across platforms.
```
I think the comment on the `#pragma` line is plenty to be put in the code.
```suggestion
```

martincostello commented on src/OpenTelemetry/Metrics/Reader/PeriodicExportingMetricReaderOptions.cs:
```diff
@@ -52,4 +52,9 @@ internal PeriodicExportingMetricReaderOptions(IConfiguration configuration)
     public int? ExportTimeoutMilliseconds { get; set; }
+
+    /// <summary>
+    /// Gets or sets a value indicating whether to use threads. Enables the use of <see cref="Thread" /> when true, <see cref="Task"/> when false. The default value is true.
```
```suggestion
    /// Gets or sets a value indicating whether to use threads. Enables the use of <see cref="Thread" /> when <see langword="true"/>; otherwise <see cref="Task"/> is used. The default value is <see langword="true"/>.
```

martincostello commented on test/OpenTelemetry.Tests/Internal/PeriodicExportingMetricReaderHelperTests.cs:
```diff
@@ -33,6 +33,18 @@ public void CreatePeriodicExportingMetricReader_Defaults()
+    [Fact]
+    public void CreatePeriodicExportingMetricReader_Defaults_WithTask()
+    {
+#pragma warning disable CA2000 // Dispose objects before losing scope
+        var reader = CreatePeriodicExportingMetricReader(useThreads: false);
+#pragma warning restore CA2000 // Dispose objects before losing scope
```
Why suppress rather than fix?

jeromelaban replied:
Mostly because the same test above is doing it, but indeed, it can be disposed.

martincostello commented on src/OpenTelemetry/Internal/PeriodicExportingMetricReaderWorker.cs:
```diff
@@ -0,0 +1,67 @@
+/// </summary>
+internal abstract class PeriodicExportingMetricReaderWorker : IDisposable
+{
+    protected readonly BaseExportingMetricReader metricReader;
+    protected readonly int exportIntervalMilliseconds;
+    protected readonly int exportTimeoutMilliseconds;
```
Properties rather than fields? Non-private fields should be PascalCase if not?

martincostello commented on src/OpenTelemetry/Internal/PeriodicExportingMetricReaderWorker.cs:
```diff
@@ -0,0 +1,67 @@
+    /// <summary>
+    /// Triggers an export operation.
+    /// </summary>
+    /// <returns>True if the export was triggered successfully; otherwise, false.</returns>
```
```suggestion
    /// <returns><see langword="true"/> if the export was triggered successfully; otherwise, <see langword="false"/>.</returns>
```

martincostello commented on src/OpenTelemetry/Internal/PeriodicExportingMetricReaderWorker.cs:
```diff
@@ -0,0 +1,67 @@
+    /// Initiates shutdown and waits for completion.
+    /// </summary>
+    /// <param name="timeoutMilliseconds">The timeout in milliseconds.</param>
+    /// <returns>True if the shutdown completed within the timeout; otherwise, false.</returns>
```
```suggestion
    /// <returns><see langword="true"/> if the shutdown completed within the timeout; otherwise, <see langword="false"/>.</returns>
```

martincostello commented on src/OpenTelemetry/Internal/PeriodicExportingMetricReaderWorker.cs:
```diff
@@ -0,0 +1,67 @@
+    /// <returns>True if the shutdown completed within the timeout; otherwise, false.</returns>
+    public abstract bool Shutdown(int timeoutMilliseconds);
+
+    /// <summary>
+    /// Disposes of the worker and its resources.
+    /// </summary>
+    public void Dispose()
+    {
+        this.Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    /// <summary>
+    /// Releases the unmanaged resources used by this class and optionally releases the managed resources.
+    /// </summary>
+    /// <param name="disposing">true to release both managed and unmanaged resources; false to release only unmanaged resources.</param>
+    protected virtual void Dispose(bool disposing)
+    {
+    }
```
Needs a finalizer for the dispose pattern.

martincostello commented on src/OpenTelemetry/BatchExportProcessor.cs:
```diff
@@ -20,16 +20,13 @@ public abstract class BatchExportProcessor<T> : BaseExportProcessor<T>
     internal const int DefaultMaxExportBatchSize = 512;
 
     internal readonly int MaxExportBatchSize;
+
```
```suggestion
```

martincostello commented on src/OpenTelemetry/BatchExportProcessor.cs:
```diff
@@ -116,113 +125,29 @@ protected override void OnExport(T data)
         }
 
         var sw = Stopwatch.StartNew();
-        this.exporterThread.Join(timeoutMilliseconds);
         var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;
```
Same comment here regarding the redundant stopwatch.

martincostello commented on src/OpenTelemetry/Internal/BatchExportTaskWorker.cs:
```diff
@@ -0,0 +1,256 @@
+    {
+        this.workerTask = Task.Factory.StartNew(
+            this.ExporterProcAsync,
+            CancellationToken.None,
```
Maybe change the signature to allow the `CancellationToken` to be passed through? We could still pass in `CancellationToken.None` if neccessary but it might make the code easier to test.

Or alternatively pass through `cancellationTokenSource.Token`.

martincostello commented on src/OpenTelemetry/Internal/BatchExportTaskWorker.cs:
```diff
@@ -0,0 +1,256 @@
+
+        // There is a chance that the export task finished processing all the data from the queue,
+        // and signaled before we enter wait here, use polling to prevent being blocked indefinitely.
+        const int pollingMilliseconds = 1000;
```
Is it possible to factor this in terms of `TimeSpan` (e.g. with `Stopwatch.Elapsed`) rather than integer milliseconds? It might make the code more readable, and I try to avoid `int` for timeouts where possible so bugs from things like mixing seconds and milliseconds don't occur.

martincostello replied:
OK no problem - I thought it might have been tricky with the various TFMs and possible missing overloads anyway, so if there's no easy way to refactor it then let's leave it in terms of `int`.

martincostello commented on src/OpenTelemetry/Internal/BatchExportThreadWorker.cs:
```diff
@@ -0,0 +1,223 @@
+
+        // There is a chance that the export thread finished processing all the data from the queue,
+        // and signaled before we enter wait here, use polling to prevent being blocked indefinitely.
+        const int pollingMilliseconds = 1000;
```
Similar question here.

martincostello commented on src/OpenTelemetry/Internal/BatchExportWorker.cs:
```diff
@@ -0,0 +1,133 @@
+internal abstract class BatchExportWorker<T> : IDisposable
+    where T : class
+{
+    protected readonly CircularBuffer<T> circularBuffer;
+    protected readonly BaseExporter<T> exporter;
+    protected readonly int maxExportBatchSize;
+    protected readonly int scheduledDelayMilliseconds;
+    protected readonly int exporterTimeoutMilliseconds;
```
Same casing vs. property comment.

martincostello commented on src/OpenTelemetry/Internal/BatchExportWorker.cs:
```diff
@@ -0,0 +1,133 @@
+        return this.circularBuffer.RemovedCount >= this.ShutdownDrainTarget;
+    }
+
+    /// <summary>
+    /// Releases the unmanaged resources used by this class and optionally releases the managed resources.
+    /// </summary>
+    /// <param name="disposing">True to release both managed and unmanaged resources; false to release only unmanaged resources.</param>
+    protected virtual void Dispose(bool disposing)
+    {
+    }
```
Add finalizer.

martincostello commented on src/OpenTelemetry/Internal/PeriodicExportingMetricReaderTaskWorker.cs:
Same comment in this file to previous ones regarding cancellation token and TimeSpan.

jeromelaban committed:
chore: Adjust after review

- Adjusts null propagation, naming, and comments.
- Adjusts BatchExportTaskWorker and PeriodicExportingMetricReaderTaskWorker to start the Task with a known cancellation token

jeromelaban commented:
Thanks for the reviews @martincostello!

verdie-g commented on src/OpenTelemetry/Internal/BatchExportTaskWorker.cs:
```diff
@@ -0,0 +1,256 @@
+        this.workerTask = Task.Factory.StartNew(
+            this.ExporterProcAsync,
+            this.cancellationTokenSource.Token,
+            TaskCreationOptions.LongRunning,
```
This option means creating a dedicated thread, defeating the purpose of this implementation. Though, since `ExporterProcAsync` can yield the execution, the continuation will be schedule on the thread pool and the thread will probably die.

A simple Task.Run would be safer here.

rajkumar-rangaraj replied:
`TaskCreationOptions.LongRunning` hints the scheduler to create a dedicated thread, which is exactly what this PR is trying to avoid on WebAssembly. As @verdie-g pointed out in the PR comments, a simple `Task.Run(this.ExporterProcAsync)` would be safer. On WASM where threads aren't available, `LongRunning` may either be ignored or throw.

verdie-g commented on src/OpenTelemetry/Internal/BatchExportTaskWorker.cs:
```diff
@@ -0,0 +1,256 @@
+                await Task.WhenAny(
+                    this.dataExportedNotification.Task,
+                    this.shutdownCompletionSource.Task,
+                    Task.Delay(timeout, combinedTokenSource.Token)).ConfigureAwait(false);
```
Since the `Task.Delay` is used as a timeout mechanism, the delay result will be discarded most of the time since the other would already have completed. Yet, we let the timer fire every time here which can cause important contention (see https://github.com/dotnet/runtime/issues/83890#issuecomment-1503158793). The fix would be to cancel the Task.Delay when it did not complete.

rajkumar-rangaraj replied:
While `combinedTokenSource` exists, it only cancels on overall shutdown/timeout, not when `dataExportedNotification `completes. Each loop iteration leaks a live timer until it fires. This was also flagged by @verdie-g. The fix: cancel the `CancellationTokenSource` after `WhenAny` returns, or restructure to reuse a single timer.

Same issue exists in `ExporterProcAsync`:

```
await Task.WhenAny(
    this.exportTrigger.WaitAsync(cancellationToken),
    Task.Delay(this.ScheduledDelayMilliseconds, cancellationToken)).ConfigureAwait(false);
```

The `Task.Delay` is only cancelled on full shutdown (`cancellationToken`), not when the semaphore fires, so timers accumulate.

madskonradsen commented:
Let's keep this alive.

jeromelaban commented:
Indeed, I'll continue working on it soon.

alanwest reviewed (commented):
I have not had a chance to do a thorough review of this PR, but we talked briefly about it in our meeting this week.

I raised a concern about the UseThread configuration (see below).

We also talked about Blazor support in general. Do we know if this is the only issue blocking us from declaring support for Blazor? If we want to declare support for Blazor it would be good to develop an integration test.

alexaka1 commented:
Bad bot

alexaka1 commented:
Bad bot

verdie-g commented on src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt:
```diff
@@ -0,0 +1,8 @@
+OpenTelemetry.BatchExportProcessorOptions<T>.UseThreads.set -> void
+OpenTelemetry.BatchLogRecordExportProcessor.BatchLogRecordExportProcessor(OpenTelemetry.BaseExporter<OpenTelemetry.Logs.LogRecord!>! exporter, OpenTelemetry.BatchExportProcessorOptions<OpenTelemetry.Logs.LogRecord!>! options) -> void
+OpenTelemetry.Metrics.PeriodicExportingMetricReader.PeriodicExportingMetricReader(OpenTelemetry.BaseExporter<OpenTelemetry.Metrics.Metric!>! exporter, OpenTelemetry.Metrics.PeriodicExportingMetricReaderOptions! options) -> void
+OpenTelemetry.Metrics.PeriodicExportingMetricReaderOptions.UseThreads.get -> bool
+OpenTelemetry.Metrics.PeriodicExportingMetricReaderOptions.UseThreads.set -> void
```
> Since it sounds like your primary concern is WebAssembly/Blazor support, would it be possible to gate using some kind of platform check?

From the discussion in https://github.com/open-telemetry/opentelemetry-dotnet/pull/5838, exposing this option would also enable exporters to use the thread pool instead of unnecessarily creating threads.

lewing commented on src/OpenTelemetry/BatchExportProcessor.cs:
```diff
@@ -243,49 +163,27 @@ protected override void Dispose(bool disposing)
-        while (true)
+#if NET
+        // Use task-based worker for browser platform where threading may be limited
+        if (OperatingSystem.IsBrowser() || !this.useThreads)
```
something like 
```
    public static bool IsThreadingEnabled()
    {
        // the threadpool isn't using threads so assume starting one will fail
        ThreadPool.GetMaxThreads(out int workerThreads, out int completionPortThreads);
        return workerThreads > 1 || completionPortThreads > 1;
    }
```
would likely only ever return false on platforms where threading doesn't work while avoiding the IsBrowser check

jeromelaban replied:
That makes sense, I'll do the change.

jeromelaban replied:
This particular comment https://github.com/open-telemetry/opentelemetry-dotnet/pull/6379#discussion_r2311386226 was about avoiding the detection of Wasm only by `IsBrowser` (Wasi can also be considered with this change), and checking for the available worker threads count can indicate that we're in a thread-less context.

The presence of the API to actually use threads or not is a different question https://github.com/open-telemetry/opentelemetry-dotnet/pull/6379#discussion_r2274687783, which remains opened for now.

verdie-g replied:
Understood

pavelsavara replied:
Public API proposal https://github.com/dotnet/runtime/issues/77541

alexaka1 commented:
Bad bot

jeromelaban committed:
chore: use IsThreadingDisabled instead of browser only

alexaka1 commented:
Silence bot. This PR is more important than your programming.

jeromelaban commented:
> In case anyone isn't following the WASM issue, WASM multi-threading has been delayed indefinitely. [dotnet/aspnetcore#17730 (comment)](https://github.com/dotnet/aspnetcore/issues/17730#issuecomment-3385570216)

Indeed, this make this PR a little more important than it already was :)

alexaka1 commented:
Bad bot

jeromelaban committed:
chore: Remove public facing UseThreads property, only rely on wasm detection

jeromelaban committed:
chore: Adjust formatting

martincostello commented on test/OpenTelemetry.Tests/Internal/PeriodicExportingMetricReaderHelperTests.cs:
```diff
@@ -33,6 +34,20 @@ public void CreatePeriodicExportingMetricReader_Defaults()
+
+        Assert.Equal(60000, reader.ExportIntervalMilliseconds);
+        Assert.Equal(30000, reader.ExportTimeoutMilliseconds);
+        Assert.Equal(MetricReaderTemporalityPreference.Cumulative, reader.TemporalityPreference);
```
```suggestion
            using var threadingOverride = ThreadingHelper.BeginThreadingOverride(threadingDisabled: true);

#pragma warning disable CA2000 // Dispose objects before losing scope
            var reader = CreatePeriodicExportingMetricReader();
#pragma warning restore CA2000 // Dispose objects before losing scope

            Assert.Equal(60000, reader.ExportIntervalMilliseconds);
            Assert.Equal(30000, reader.ExportTimeoutMilliseconds);
            Assert.Equal(MetricReaderTemporalityPreference.Cumulative, reader.TemporalityPreference);
```

jeromelaban replied:
One of the target frameworks does not allow for `using`, we need to keep this unfortunately.

jeromelaban replied:
I misread, my apologies :) Will fix.

alanwest commented on src/OpenTelemetry/BatchExportProcessor.cs:
```diff
@@ -46,31 +40,48 @@ protected BatchExportProcessor(
+    /// <param name="options">Configuration options for the batch export processor.</param>
+    protected BatchExportProcessor(
+        BaseExporter<T> exporter,
+        BatchExportProcessorOptions<T> options)
```
Why is this new constructor overload necessary?

rajkumar-rangaraj commented:
@jeromelaban Could you please update the PR description? The current summary references old implementation details that no longer match the code.

Also, could you clarify the rationale for the new public constructors:

```
OpenTelemetry.BatchActivityExportProcessor.BatchActivityExportProcessor(OpenTelemetry.BaseExporter<System.Diagnostics.Activity!>! exporter, OpenTelemetry.BatchExportProcessorOptions<System.Diagnostics.Activity!>! options) -> void
OpenTelemetry.BatchExportProcessor<T>.BatchExportProcessor(OpenTelemetry.BaseExporter<T!>! exporter, OpenTelemetry.BatchExportProcessorOptions<T!>! options) -> void
OpenTelemetry.BatchLogRecordExportProcessor.BatchLogRecordExportProcessor(OpenTelemetry.BaseExporter<OpenTelemetry.Logs.LogRecord!>! exporter, OpenTelemetry.BatchExportProcessorOptions<OpenTelemetry.Logs.LogRecord!>! options) -> void
OpenTelemetry.Metrics.PeriodicExportingMetricReader.PeriodicExportingMetricReader(OpenTelemetry.BaseExporter<OpenTelemetry.Metrics.Metric!>! exporter, OpenTelemetry.Metrics.PeriodicExportingMetricReaderOptions! options) -> void
```

The refactoring itself could work with these as `internal` (using `InternalsVisibleTo` for tests). Is making them public intentional for user-facing extensibility, or would internal visibility be preferred to minimize API surface?

jeromelaban committed:
chore: Adjust for incorrect double-wait time

jeromelaban committed:
chore: Remove unused public overloads

noahfalk commented on src/OpenTelemetry/Internal/ThreadingHelper.cs:
```diff
@@ -0,0 +1,65 @@
+        // if the threadpool isn't using threads assume they aren't enabled
+        ThreadPool.GetMaxThreads(out int workerThreads, out int completionPortThreads);
+
+        return workerThreads == 1 && completionPortThreads == 1;
```
You might get false positives this way. The number of threads could be 1 at the moment you check but adding new threads would work if you tried it.

The way .NET does it is checking [OperatingSystem.IsBrowser](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Metrics/AggregationManager.cs,180). We assume that true means you only have 1 thread and false means you have more. If you need a check its a reasonable one, though I won't be surprised at all if it needs to get adjusted in the future. Having the code isolated as you have it here should make those adjustments easy to do.

Given that you'd like to create a polling thread assuming they are supported, the most robust route is probably to try creating the thread and then do a non-thread approach if the creation fails. Something like:

```C#
try
{
    Thread pollingThread = new Thread(...);
    pollingThread.Start();
}
catch(PlatformNotSupportedException)
{
    // fallback code that doesn't use thread
    StartPollingTimer();
}
```

If you need a prediction about whether new threads will be available before you try to create one this won't give you that capability, but it didn't appear that was necessary for this PR.

Hope that helps!

jeromelaban committed:
chore: fix documentation and unused namespace

jeromelaban committed:
chore: Adjust task creation and cancellation

jeromelaban commented on src/OpenTelemetry/Metrics/Reader/PeriodicExportingMetricReader.cs:
```diff
@@ -47,44 +45,43 @@ public PeriodicExportingMetricReader(
         }
-        catch (ObjectDisposedException)
+
+        var result = this.worker.Shutdown(timeoutMilliseconds);
```
The ObjectDisposedException is tracking is removed, but handled through shutdown, non-zero timeouts behave the same way.

For timeout 0, previously the join would almost always return false, meaning that result was false. Now, there's no join and so the result only depends on exporter.Shutdown(). In practice, timeout=0 means don't wait but still report failures.

jeromelaban committed:
chore: Adjust for net7 and below

jeromelaban committed:
chore: Adjust formatting

jeromelaban committed:
chore: Adjust namespace

rajkumar-rangaraj commented:
@jeromelaban Thanks for the continued work on this! A couple of items still pending before we can move forward:

1. **`.GetAwaiter().GetResult()` deadlock risk ([thread](https://github.com/open-telemetry/opentelemetry-dotnet/pull/6379#discussion_r2849745780)):** You acknowledged this can deadlock on Blazor's `SynchronizationContext`. Could you implement one of the options you proposed?

2. **Merge conflicts:** The PR has conflicts with `main` that need to be resolved.

3. **PR description:** Could you update it to reflect the current implementation (no public `useThreads` API, automatic detection via `ThreadingHelper`)?

jeromelaban committed:
chore: Handle export triggering for Wasm environments to prevent deadlocks

jeromelaban committed:
chore: Remove cancellation token from Task.Run in batch export and periodic exporting workers

Lets the task cancel itself through existing flows.

rajkumar-rangaraj committed:
Merge branch 'main' into dev/jela/tasks

rajkumar-rangaraj reviewed (approved):
LGTM, left minor feedback on dispose pattern.

jeromelaban committed:
chore: Refactor Dispose method in BatchExportTaskWorker to improve resource management

rajkumar-rangaraj committed:
Merge branch 'main' into dev/jela/tasks
